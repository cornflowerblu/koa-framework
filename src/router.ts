import Koa from 'koa'
import Router from 'koa-router'
import invariant from 'tiny-invariant'
import { createNewUser } from './controllers/user'
import {
  getProtectedHomepage,
  getUnprotectedHomepage,
} from './controllers/homepage'

type Routes = {
  name: string
  path: string
}

export const getRoute = (name: string): string => {
  let routeName = routes.find((el) => el.name === name)
  invariant(routeName)
  return routeName?.path
}

export const router = new Router()

const routes: Array<Routes> = [
  {
    name: 'home page protected',
    path: '/',
  },
  {
    name: 'landing page un-protected',
    path: '/page',
  },
  {
    name: 'create new user',
    path: '/user/create',
  },
]

router.get(getRoute('home page protected'), (ctx: Koa.Context) =>
  getProtectedHomepage(ctx)
)

router.get(getRoute('landing page un-protected'), (ctx: Koa.Context) =>
  getUnprotectedHomepage(ctx)
)

router.post(getRoute('create new user'), (ctx: Koa.Context) =>
  createNewUser(ctx)
)

export default routes
