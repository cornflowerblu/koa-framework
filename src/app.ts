import Koa from 'koa'
import json from 'koa-json'
import { router } from './router'
import bodyParser from 'koa-bodyparser'

const app = new Koa()

app.use(json())
app.use(bodyParser())
app.use(router.routes())

export default app
