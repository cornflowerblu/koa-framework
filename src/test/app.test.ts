import request from 'supertest'
import app from '../app'
import { getRoute } from '../router'

test('Unauthorized page works', async () => {
  const url = getRoute('home page protected')
  const response = await request(app.callback()).get(url)
  expect(response.status).toBe(403)
  expect(response.text).toMatchSnapshot()
})

test('landing page un-protected works', async () => {
  const url = getRoute('landing page un-protected')
  const response = await request(app.callback()).get(url)
  expect(response.status).toBe(200)
  expect(response.text).toMatchSnapshot()
})
