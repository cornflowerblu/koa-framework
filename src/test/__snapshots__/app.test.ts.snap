// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unauthorized page works 1`] = `
"{
  \\"statusCode\\": 403,
  \\"error\\": \\"Forbidden\\",
  \\"message\\": \\"Forbidden\\"
}"
`;

exports[`landing page un-protected works 1`] = `"Hi!"`;
